### ISSUE TITLE
Transformer - RNN-Transducer: train / validation loss gets worse after 20 epochs

### ISSUE THREAD

**Issue by francampillo** [espnet/espnet#2700]
Hi all,

I'm trying to train transformer - RNN_transducer acoustic models for German using the Vivos recipe as reference, and even though everything seems to go ok in the first 20 epochs, after that the loss function gets noticeably worse both for the training and validation sets, as shown in the figure:

![loss](https://user-images.githubusercontent.com/51226014/99687572-1b2ec900-2a85-11eb-93cc-d3c4105fd52a.png)

The training data seems to be right, since I could train with it a hybrid CTC / Attention based transformer model with a very good performance.

I have also observed that the gradient norm begins to change drastically before that epoch:

![grad_norm](https://user-images.githubusercontent.com/51226014/99688328-f2f39a00-2a85-11eb-8333-58598decff0c.jpg)

If I'm getting it right, these huge gradient norms will result into tiny clipping coefficients (in clip_grad_norm_), and that will affect negatively the parameter gradients.

I've checked this issue [2154](https://github.com/espnet/espnet/issues/2154), but it doesn't look exactly like the same case, since the problem there seems to be already at the beginning of training, whereas here that part looks ok. [2652](https://github.com/espnet/espnet/issues/2652) looks very similar, but it's for a Conformer model, and with a smaller training database.

Could anyone please help me with this? Reducing the learning rate might fix the problem in this specific case, but I wonder whether we can learn some more from this, and avoid it in future experiments.

Some context data:

- ESPnet: commit 2a74c31cd6fcd8192d2498fc98e6dd6486fe0e43
- CUDA: 10.0
- python: 3.8.3
- pytorch: 1.4.0.
- Amount of training data: ~1300 hours.

The configuration file for the call to asr_train.py is:

    backend: pytorch
    resume: ''
    seed: 1
    verbose: 1
    etype: transformer
    transformer-enc-input-layer: vgg2l
    enc-block-arch:
    - type: transformer
        d_hidden: 512
        d_ff: 2048
        heads: 8
        att-dropout-rate: 0.0
        dropout-rate: 0.1
    enc-block-repeat: 12
    dtype: lstm
    dlayers: 2
    dec-embed-dim: 700
    dunits: 700
    dropout-rate-decoder: 0.2
    dropout-rate-embed-decoder: 0.1
    joint-dim: 640
    transformer-lr: 10.0
    transformer-warmup-steps: 25000
    rnnt-mode: rnnt
    model-module: espnet.nets.pytorch_backend.e2e_asr_transducer:E2E
    batch-bins: 8000000
    minibatches: 0
    maxlen-in: 512
    maxlen-out: 150
    criterion: loss
    early-stop-criterion: validation/main/loss
    sortagrad: 0
    opt: noam
    accum-grad: 4
    grad-clip: 5
    patience: 0
    epochs: 50

Thank you very much in advance, both for this great framework and your help :)

**Comment by b-flo** [comment]
Hi,

> Could anyone please help me with this? Reducing the learning rate might fix the problem in this specific case

Yes, adjusting `transformer-lr` and `transformer-warmup-steps` may fix the problem. `batch-bins` may also be modified but I rarely use it, I would suggest following already given advices for this one (in posts you linked or other ones).

>  but I wonder whether we can learn some more from this, and avoid it in future experiments.

For sure, I'm aware of the behaviour and the impact on performances with T-T but I couldn't investigate thoroughly because the TO-DO was piling up... After my current items are finished we will work on adding configs for recipes and we'll address such cases. Stay tuned I guess!

**Comment by francampillo** [comment]
Thank you very much for your answer, @b-flo :) Yes, I'll try that.

I'll also try to find time to investigate on this issue. Perhaps some control on the values of clif_coef in clip_grad_norm_ (similar to what is done now checking that the value is < 1) before updating the parameter gradients is enough.

Thanks again!

**Comment by b-flo** [comment]
> I'll also try to find time to investigate on this issue. 

Thank you, user feedback is really appreciate! It's easier for me to have a starting point as resources and time are allocated on other things :)

> Perhaps some control on the values of clif_coef in clip_grad_norm_ (similar to what is done now checking that the value is < 1) before updating the parameter gradients is enough.

It may be worth checking that part given your case but I'm suspecting other training issues I need to investigate before giving a firm opinion.

I'm closing this issue for now but feel free to re-open or open a new one as discussion when you've investigated!
